# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/angle/checkout/src/libANGLE/renderer/load_functions_table_autogen.cpp
# Commit: 3a91885d47e0
# Full Hash: 3a91885d47e0ef7fba9749562d52300eea67c651
# Author: Miko Mynttinen <mikokm@gmail.com>
# Date: 2019-11-11 21:52:52
# Regressor Bug: 1578576
# File Overlap Count: 4
# Description:
#   Bug 1578576 - Part 2: Update to ANGLE 3865 r=jgilbert
#   
#   Depends on D44561
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D44579
# ==============================================================================

diff -r 77cee77c0307 -r 3a91885d47e0 gfx/angle/checkout/src/libANGLE/renderer/load_functions_table_autogen.cpp
--- a/gfx/angle/checkout/src/libANGLE/renderer/load_functions_table_autogen.cpp	Fri Nov 08 16:41:42 2019 +0000
+++ b/gfx/angle/checkout/src/libANGLE/renderer/load_functions_table_autogen.cpp	Fri Nov 08 17:06:00 2019 +0000
@@ -1,7 +1,7 @@
 // GENERATED FILE - DO NOT EDIT.
 // Generated by gen_load_functions_table.py using data from load_functions_data.json
 //
-// Copyright 2018 The ANGLE Project Authors. All rights reserved.
+// Copyright 2019 The ANGLE Project Authors. All rights reserved.
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 //
@@ -56,6 +56,20 @@
     UNREACHABLE();
 }
 
+LoadImageFunctionInfo A1RGB5_ANGLEX_to_A1R5G5B5_UNORM(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadRGB5A1ToRGBA8, true);
+        case GL_UNSIGNED_SHORT_1_5_5_5_REV_EXT:
+            return LoadImageFunctionInfo(LoadToNative<GLushort, 1>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo ALPHA_to_R16G16B16A16_FLOAT(GLenum type)
 {
     switch (type)
@@ -165,7 +179,7 @@
     switch (type)
     {
         case GL_UNSIGNED_BYTE:
-            return LoadImageFunctionInfo(LoadToNative<GLushort, 1>, false);
+            return LoadImageFunctionInfo(LoadRGB8ToBGR565, true);
         case GL_UNSIGNED_SHORT_5_6_5:
             return LoadImageFunctionInfo(LoadRGB565ToBGR565, true);
         default:
@@ -179,9 +193,9 @@
     switch (type)
     {
         case GL_UNSIGNED_BYTE:
-            return LoadImageFunctionInfo(LoadToNative<GLubyte, 4>, false);
+            return LoadImageFunctionInfo(LoadRGB5A1ToRGBA8, true);
         case GL_UNSIGNED_SHORT_1_5_5_5_REV_EXT:
-            return LoadImageFunctionInfo(LoadRGB5A1ToRGBA8, true);
+            return LoadImageFunctionInfo(LoadToNative<GLushort, 1>, false);
         default:
             UNREACHABLE();
             return LoadImageFunctionInfo(UnreachableLoadFunction, true);
@@ -214,6 +228,18 @@
     }
 }
 
+LoadImageFunctionInfo BGRA8_SRGB_ANGLEX_to_B8G8R8A8_UNORM_SRGB(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadToNative<GLubyte, 4>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo BGRA_EXT_to_default(GLenum type)
 {
     switch (type)
@@ -226,6 +252,42 @@
     }
 }
 
+LoadImageFunctionInfo BGRX8_ANGLEX_to_B8G8R8A8_UNORM(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadToNative<GLubyte, 4>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo BGRX8_ANGLEX_to_B8G8R8X8_UNORM(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadToNative<GLubyte, 4>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_R11_EAC_to_EAC_R11_UNORM_BLOCK(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 8>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo COMPRESSED_R11_EAC_to_R16_UNORM(GLenum type)
 {
     switch (type)
@@ -238,6 +300,18 @@
     }
 }
 
+LoadImageFunctionInfo COMPRESSED_RG11_EAC_to_EAC_R11G11_UNORM_BLOCK(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo COMPRESSED_RG11_EAC_to_R16G16_UNORM(GLenum type)
 {
     switch (type)
@@ -250,6 +324,18 @@
     }
 }
 
+LoadImageFunctionInfo COMPRESSED_RGB8_ETC2_to_ETC2_R8G8B8_UNORM_BLOCK(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 8>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo COMPRESSED_RGB8_ETC2_to_R8G8B8A8_UNORM(GLenum type)
 {
     switch (type)
@@ -274,6 +360,19 @@
     }
 }
 
+LoadImageFunctionInfo COMPRESSED_RGB8_PUNCHTHROUGH_ALPHA1_ETC2_to_ETC2_R8G8B8A1_UNORM_BLOCK(
+    GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 8>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo COMPRESSED_RGB8_PUNCHTHROUGH_ALPHA1_ETC2_to_R8G8B8A8_UNORM(GLenum type)
 {
     switch (type)
@@ -299,6 +398,18 @@
     }
 }
 
+LoadImageFunctionInfo COMPRESSED_RGBA8_ETC2_EAC_to_ETC2_R8G8B8A8_UNORM_BLOCK(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo COMPRESSED_RGBA8_ETC2_EAC_to_R8G8B8A8_UNORM(GLenum type)
 {
     switch (type)
@@ -311,12 +422,300 @@
     }
 }
 
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_10x10_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<10, 10, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_10x5_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<10, 5, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_10x6_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<10, 6, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_10x8_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<10, 8, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_12x10_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<12, 10, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_12x12_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<12, 12, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_3x3x3_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<3, 3, 3, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_4x3x3_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 3, 3, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_4x4_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_4x4x3_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 3, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_4x4x4_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 4, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_5x4_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<5, 4, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_5x4x4_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<5, 4, 4, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_5x5_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<5, 5, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_5x5x4_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<5, 5, 4, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_5x5x5_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<5, 5, 5, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_6x5_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<6, 5, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_6x5x5_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<6, 5, 5, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_6x6_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<6, 6, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_6x6x5_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<6, 6, 5, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_6x6x6_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<6, 6, 6, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_8x5_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<8, 5, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_8x6_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<8, 6, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_RGBA_ASTC_8x8_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<8, 8, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo COMPRESSED_RGBA_BPTC_UNORM_EXT_to_default(GLenum type)
 {
     switch (type)
     {
         case GL_UNSIGNED_BYTE:
-            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 16>, true);
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 16>, true);
         default:
             UNREACHABLE();
             return LoadImageFunctionInfo(UnreachableLoadFunction, true);
@@ -328,7 +727,7 @@
     switch (type)
     {
         case GL_UNSIGNED_BYTE:
-            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 8>, true);
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 8>, true);
         default:
             UNREACHABLE();
             return LoadImageFunctionInfo(UnreachableLoadFunction, true);
@@ -340,7 +739,7 @@
     switch (type)
     {
         case GL_UNSIGNED_BYTE:
-            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 16>, true);
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 16>, true);
         default:
             UNREACHABLE();
             return LoadImageFunctionInfo(UnreachableLoadFunction, true);
@@ -352,7 +751,7 @@
     switch (type)
     {
         case GL_UNSIGNED_BYTE:
-            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 16>, true);
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 16>, true);
         default:
             UNREACHABLE();
             return LoadImageFunctionInfo(UnreachableLoadFunction, true);
@@ -364,7 +763,7 @@
     switch (type)
     {
         case GL_UNSIGNED_BYTE:
-            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 16>, true);
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 16>, true);
         default:
             UNREACHABLE();
             return LoadImageFunctionInfo(UnreachableLoadFunction, true);
@@ -376,7 +775,7 @@
     switch (type)
     {
         case GL_UNSIGNED_BYTE:
-            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 16>, true);
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 16>, true);
         default:
             UNREACHABLE();
             return LoadImageFunctionInfo(UnreachableLoadFunction, true);
@@ -388,7 +787,19 @@
     switch (type)
     {
         case GL_UNSIGNED_BYTE:
-            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 8>, true);
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 8>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SIGNED_R11_EAC_to_EAC_R11_SNORM_BLOCK(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 8>, true);
         default:
             UNREACHABLE();
             return LoadImageFunctionInfo(UnreachableLoadFunction, true);
@@ -407,6 +818,18 @@
     }
 }
 
+LoadImageFunctionInfo COMPRESSED_SIGNED_RG11_EAC_to_EAC_R11G11_SNORM_BLOCK(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo COMPRESSED_SIGNED_RG11_EAC_to_R16G16_SNORM(GLenum type)
 {
     switch (type)
@@ -419,6 +842,306 @@
     }
 }
 
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_10x10_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<10, 10, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_10x5_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<10, 5, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_10x6_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<10, 6, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_10x8_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<10, 8, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_12x10_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<12, 10, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_12x12_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<12, 12, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_3x3x3_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<3, 3, 3, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_4x3x3_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 3, 3, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_4x4_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_4x4x3_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 3, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_4x4x4_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 4, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_5x4_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<5, 4, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_5x4x4_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<5, 4, 4, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_5x5_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<5, 5, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_5x5x4_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<5, 5, 4, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_5x5x5_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<5, 5, 5, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_6x5_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<6, 5, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_6x5x5_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<6, 5, 5, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_6x6_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<6, 6, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_6x6x5_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<6, 6, 5, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_6x6x6_OES_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<6, 6, 6, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_8x5_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<8, 5, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_8x6_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<8, 6, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ASTC_8x8_KHR_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<8, 8, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ETC2_EAC_to_ETC2_R8G8B8A8_SRGB_BLOCK(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 16>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo COMPRESSED_SRGB8_ALPHA8_ETC2_EAC_to_R8G8B8A8_UNORM_SRGB(GLenum type)
 {
     switch (type)
@@ -431,6 +1154,18 @@
     }
 }
 
+LoadImageFunctionInfo COMPRESSED_SRGB8_ETC2_to_ETC2_R8G8B8_SRGB_BLOCK(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 8>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo COMPRESSED_SRGB8_ETC2_to_R8G8B8A8_UNORM_SRGB(GLenum type)
 {
     switch (type)
@@ -456,6 +1191,19 @@
     }
 }
 
+LoadImageFunctionInfo COMPRESSED_SRGB8_PUNCHTHROUGH_ALPHA1_ETC2_to_ETC2_R8G8B8A1_SRGB_BLOCK(
+    GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 8>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo COMPRESSED_SRGB8_PUNCHTHROUGH_ALPHA1_ETC2_to_R8G8B8A8_UNORM_SRGB(GLenum type)
 {
     switch (type)
@@ -487,7 +1235,7 @@
     switch (type)
     {
         case GL_UNSIGNED_BYTE:
-            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 16>, true);
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 16>, true);
         default:
             UNREACHABLE();
             return LoadImageFunctionInfo(UnreachableLoadFunction, true);
@@ -499,7 +1247,7 @@
     switch (type)
     {
         case GL_UNSIGNED_BYTE:
-            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 8>, true);
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 8>, true);
         default:
             UNREACHABLE();
             return LoadImageFunctionInfo(UnreachableLoadFunction, true);
@@ -511,7 +1259,7 @@
     switch (type)
     {
         case GL_UNSIGNED_BYTE:
-            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 16>, true);
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 16>, true);
         default:
             UNREACHABLE();
             return LoadImageFunctionInfo(UnreachableLoadFunction, true);
@@ -523,7 +1271,7 @@
     switch (type)
     {
         case GL_UNSIGNED_BYTE:
-            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 16>, true);
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 16>, true);
         default:
             UNREACHABLE();
             return LoadImageFunctionInfo(UnreachableLoadFunction, true);
@@ -535,7 +1283,7 @@
     switch (type)
     {
         case GL_UNSIGNED_BYTE:
-            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 8>, true);
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 8>, true);
         default:
             UNREACHABLE();
             return LoadImageFunctionInfo(UnreachableLoadFunction, true);
@@ -554,6 +1302,30 @@
     }
 }
 
+LoadImageFunctionInfo DEPTH24_STENCIL8_to_D32_FLOAT_S8X24_UINT(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_INT_24_8:
+            return LoadImageFunctionInfo(LoadD24S8ToD32FS8X24, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo DEPTH32F_STENCIL8_to_D24_UNORM_S8_UINT(GLenum type)
+{
+    switch (type)
+    {
+        case GL_FLOAT_32_UNSIGNED_INT_24_8_REV:
+            return LoadImageFunctionInfo(LoadD32FS8X24ToD24S8, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo DEPTH32F_STENCIL8_to_D32_FLOAT_S8X24_UINT(GLenum type)
 {
     switch (type)
@@ -592,6 +1364,18 @@
     }
 }
 
+LoadImageFunctionInfo DEPTH_COMPONENT24_to_D32_FLOAT_S8X24_UINT(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_INT_24_8:
+            return LoadImageFunctionInfo(LoadD24S8ToD32FS8X24, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo DEPTH_COMPONENT32F_to_D32_FLOAT(GLenum type)
 {
     switch (type)
@@ -628,6 +1412,30 @@
     }
 }
 
+LoadImageFunctionInfo ETC1_RGB8_OES_to_ETC1_R8G8B8_UNORM_BLOCK(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 8>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo ETC1_RGB8_OES_to_ETC2_R8G8B8_UNORM_BLOCK(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadCompressedToNative<4, 4, 1, 8>, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo ETC1_RGB8_OES_to_R8G8B8A8_UNORM(GLenum type)
 {
     switch (type)
@@ -906,6 +1714,30 @@
     }
 }
 
+LoadImageFunctionInfo R16_SSCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo R16_USCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo R32F_to_R32_FLOAT(GLenum type)
 {
     switch (type)
@@ -990,6 +1822,30 @@
     }
 }
 
+LoadImageFunctionInfo R8_SSCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo R8_USCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo RG16F_to_R16G16_FLOAT(GLenum type)
 {
     switch (type)
@@ -1054,6 +1910,30 @@
     }
 }
 
+LoadImageFunctionInfo RG16_SSCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo RG16_USCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo RG32F_to_R32G32_FLOAT(GLenum type)
 {
     switch (type)
@@ -1138,6 +2018,30 @@
     }
 }
 
+LoadImageFunctionInfo RG8_SSCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo RG8_USCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo RGB_to_default(GLenum type)
 {
     switch (type)
@@ -1176,6 +2080,58 @@
     }
 }
 
+LoadImageFunctionInfo RGB10_A2_SINT_ANGLEX_to_R10G10B10A2_SINT(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadRGB10A2ToRGBA8, true);
+        case GL_UNSIGNED_INT_2_10_10_10_REV:
+            return LoadImageFunctionInfo(LoadToNative<GLuint, 1>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo RGB10_A2_SNORM_ANGLEX_to_R10G10B10A2_SNORM(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadRGB10A2ToRGBA8, true);
+        case GL_UNSIGNED_INT_2_10_10_10_REV:
+            return LoadImageFunctionInfo(LoadToNative<GLuint, 1>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo RGB10_A2_SSCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo RGB10_A2_USCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo RGB16F_to_R16G16B16A16_FLOAT(GLenum type)
 {
     switch (type)
@@ -1192,6 +2148,22 @@
     }
 }
 
+LoadImageFunctionInfo RGB16F_to_R16G16B16_FLOAT(GLenum type)
+{
+    switch (type)
+    {
+        case GL_FLOAT:
+            return LoadImageFunctionInfo(LoadRGB32FToRGB16F, true);
+        case GL_HALF_FLOAT:
+            return LoadImageFunctionInfo(LoadToNative<GLhalf, 3>, false);
+        case GL_HALF_FLOAT_OES:
+            return LoadImageFunctionInfo(LoadToNative<GLhalf, 3>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo RGB16I_to_R16G16B16A16_SINT(GLenum type)
 {
     switch (type)
@@ -1204,6 +2176,18 @@
     }
 }
 
+LoadImageFunctionInfo RGB16I_to_R16G16B16_SINT(GLenum type)
+{
+    switch (type)
+    {
+        case GL_SHORT:
+            return LoadImageFunctionInfo(LoadToNative<GLshort, 3>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo RGB16UI_to_R16G16B16A16_UINT(GLenum type)
 {
     switch (type)
@@ -1216,6 +2200,18 @@
     }
 }
 
+LoadImageFunctionInfo RGB16UI_to_R16G16B16_UINT(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_SHORT:
+            return LoadImageFunctionInfo(LoadToNative<GLushort, 3>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo RGB16_EXT_to_R16G16B16A16_UNORM(GLenum type)
 {
     switch (type)
@@ -1228,6 +2224,18 @@
     }
 }
 
+LoadImageFunctionInfo RGB16_EXT_to_R16G16B16_UNORM(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_SHORT:
+            return LoadImageFunctionInfo(LoadToNative<GLushort, 3>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo RGB16_SNORM_EXT_to_R16G16B16A16_SNORM(GLenum type)
 {
     switch (type)
@@ -1240,6 +2248,42 @@
     }
 }
 
+LoadImageFunctionInfo RGB16_SNORM_EXT_to_R16G16B16_SNORM(GLenum type)
+{
+    switch (type)
+    {
+        case GL_SHORT:
+            return LoadImageFunctionInfo(LoadToNative<GLushort, 3>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo RGB16_SSCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo RGB16_USCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo RGB32F_to_R32G32B32A32_FLOAT(GLenum type)
 {
     switch (type)
@@ -1252,6 +2296,18 @@
     }
 }
 
+LoadImageFunctionInfo RGB32F_to_R32G32B32_FLOAT(GLenum type)
+{
+    switch (type)
+    {
+        case GL_FLOAT:
+            return LoadImageFunctionInfo(LoadToNative<GLfloat, 3>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo RGB32I_to_R32G32B32A32_SINT(GLenum type)
 {
     switch (type)
@@ -1264,6 +2320,18 @@
     }
 }
 
+LoadImageFunctionInfo RGB32I_to_R32G32B32_SINT(GLenum type)
+{
+    switch (type)
+    {
+        case GL_INT:
+            return LoadImageFunctionInfo(LoadToNative<GLint, 3>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo RGB32UI_to_R32G32B32A32_UINT(GLenum type)
 {
     switch (type)
@@ -1276,6 +2344,18 @@
     }
 }
 
+LoadImageFunctionInfo RGB32UI_to_R32G32B32_UINT(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_INT:
+            return LoadImageFunctionInfo(LoadToNative<GLuint, 3>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo RGB565_to_B5G6R5_UNORM(GLenum type)
 {
     switch (type)
@@ -1294,6 +2374,8 @@
 {
     switch (type)
     {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadRGB8ToBGR565, true);
         case GL_UNSIGNED_SHORT_5_6_5:
             return LoadImageFunctionInfo(LoadToNative<GLushort, 1>, false);
         default:
@@ -1320,6 +2402,10 @@
 {
     switch (type)
     {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadRGBA8ToBGR5A1, true);
+        case GL_UNSIGNED_INT_2_10_10_10_REV:
+            return LoadImageFunctionInfo(LoadRGB10A2ToBGR5A1, true);
         case GL_UNSIGNED_SHORT_5_5_5_1:
             return LoadImageFunctionInfo(LoadRGB5A1ToA1RGB5, true);
         default:
@@ -1384,6 +2470,18 @@
     }
 }
 
+LoadImageFunctionInfo RGB8I_to_R8G8B8_SINT(GLenum type)
+{
+    switch (type)
+    {
+        case GL_BYTE:
+            return LoadImageFunctionInfo(LoadToNative<GLbyte, 3>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo RGB8UI_to_R8G8B8A8_UINT(GLenum type)
 {
     switch (type)
@@ -1396,6 +2494,18 @@
     }
 }
 
+LoadImageFunctionInfo RGB8UI_to_R8G8B8_UINT(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(LoadToNative<GLubyte, 3>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo RGB8_SNORM_to_R8G8B8A8_SNORM(GLenum type)
 {
     switch (type)
@@ -1408,6 +2518,42 @@
     }
 }
 
+LoadImageFunctionInfo RGB8_SNORM_to_R8G8B8_SNORM(GLenum type)
+{
+    switch (type)
+    {
+        case GL_BYTE:
+            return LoadImageFunctionInfo(LoadToNative<GLbyte, 3>, false);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo RGB8_SSCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo RGB8_USCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo RGB9_E5_to_R9G9B9E5_SHAREDEXP(GLenum type)
 {
     switch (type)
@@ -1506,6 +2652,30 @@
     }
 }
 
+LoadImageFunctionInfo RGBA16_SSCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo RGBA16_USCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo RGBA32F_to_R32G32B32A32_FLOAT(GLenum type)
 {
     switch (type)
@@ -1618,6 +2788,30 @@
     }
 }
 
+LoadImageFunctionInfo RGBA8_SSCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
+LoadImageFunctionInfo RGBA8_USCALED_ANGLEX_to_default(GLenum type)
+{
+    switch (type)
+    {
+        case GL_UNSIGNED_BYTE:
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+        default:
+            UNREACHABLE();
+            return LoadImageFunctionInfo(UnreachableLoadFunction, true);
+    }
+}
+
 LoadImageFunctionInfo SRGB8_to_R8G8B8A8_UNORM_SRGB(GLenum type)
 {
     switch (type)
@@ -1661,6 +2855,17 @@
     // clang-format off
     switch (internalFormat)
     {
+        case GL_A1RGB5_ANGLEX:
+        {
+            switch (angleFormat)
+            {
+                case FormatID::A1R5G5B5_UNORM:
+                    return A1RGB5_ANGLEX_to_A1R5G5B5_UNORM;
+                default:
+                    break;
+            }
+            break;
+        }
         case GL_ALPHA:
         {
             switch (angleFormat)
@@ -1709,12 +2914,38 @@
             return BGRA4_ANGLEX_to_default;
         case GL_BGRA8_EXT:
             return BGRA8_EXT_to_default;
+        case GL_BGRA8_SRGB_ANGLEX:
+        {
+            switch (angleFormat)
+            {
+                case FormatID::B8G8R8A8_UNORM_SRGB:
+                    return BGRA8_SRGB_ANGLEX_to_B8G8R8A8_UNORM_SRGB;
+                default:
+                    break;
+            }
+            break;
+        }
         case GL_BGRA_EXT:
             return BGRA_EXT_to_default;
+        case GL_BGRX8_ANGLEX:
+        {
+            switch (angleFormat)
+            {
+                case FormatID::B8G8R8A8_UNORM:
+                    return BGRX8_ANGLEX_to_B8G8R8A8_UNORM;
+                case FormatID::B8G8R8X8_UNORM:
+                    return BGRX8_ANGLEX_to_B8G8R8X8_UNORM;
+                default:
+                    break;
+            }
+            break;
+        }
         case GL_COMPRESSED_R11_EAC:
         {
             switch (angleFormat)
             {
+                case FormatID::EAC_R11_UNORM_BLOCK:
+                    return COMPRESSED_R11_EAC_to_EAC_R11_UNORM_BLOCK;
                 case FormatID::R16_UNORM:
                     return COMPRESSED_R11_EAC_to_R16_UNORM;
                 default:
@@ -1726,6 +2957,8 @@
         {
             switch (angleFormat)
             {
+                case FormatID::EAC_R11G11_UNORM_BLOCK:
+                    return COMPRESSED_RG11_EAC_to_EAC_R11G11_UNORM_BLOCK;
                 case FormatID::R16G16_UNORM:
                     return COMPRESSED_RG11_EAC_to_R16G16_UNORM;
                 default:
@@ -1737,6 +2970,8 @@
         {
             switch (angleFormat)
             {
+                case FormatID::ETC2_R8G8B8_UNORM_BLOCK:
+                    return COMPRESSED_RGB8_ETC2_to_ETC2_R8G8B8_UNORM_BLOCK;
                 case FormatID::R8G8B8A8_UNORM:
                     return COMPRESSED_RGB8_ETC2_to_R8G8B8A8_UNORM;
                 default:
@@ -1759,6 +2994,8 @@
         {
             switch (angleFormat)
             {
+                case FormatID::ETC2_R8G8B8A1_UNORM_BLOCK:
+                    return COMPRESSED_RGB8_PUNCHTHROUGH_ALPHA1_ETC2_to_ETC2_R8G8B8A1_UNORM_BLOCK;
                 case FormatID::R8G8B8A8_UNORM:
                     return COMPRESSED_RGB8_PUNCHTHROUGH_ALPHA1_ETC2_to_R8G8B8A8_UNORM;
                 default:
@@ -1781,6 +3018,8 @@
         {
             switch (angleFormat)
             {
+                case FormatID::ETC2_R8G8B8A8_UNORM_BLOCK:
+                    return COMPRESSED_RGBA8_ETC2_EAC_to_ETC2_R8G8B8A8_UNORM_BLOCK;
                 case FormatID::R8G8B8A8_UNORM:
                     return COMPRESSED_RGBA8_ETC2_EAC_to_R8G8B8A8_UNORM;
                 default:
@@ -1788,6 +3027,54 @@
             }
             break;
         }
+        case GL_COMPRESSED_RGBA_ASTC_10x10_KHR:
+            return COMPRESSED_RGBA_ASTC_10x10_KHR_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_10x5_KHR:
+            return COMPRESSED_RGBA_ASTC_10x5_KHR_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_10x6_KHR:
+            return COMPRESSED_RGBA_ASTC_10x6_KHR_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_10x8_KHR:
+            return COMPRESSED_RGBA_ASTC_10x8_KHR_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_12x10_KHR:
+            return COMPRESSED_RGBA_ASTC_12x10_KHR_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_12x12_KHR:
+            return COMPRESSED_RGBA_ASTC_12x12_KHR_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_3x3x3_OES:
+            return COMPRESSED_RGBA_ASTC_3x3x3_OES_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_4x3x3_OES:
+            return COMPRESSED_RGBA_ASTC_4x3x3_OES_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_4x4_KHR:
+            return COMPRESSED_RGBA_ASTC_4x4_KHR_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_4x4x3_OES:
+            return COMPRESSED_RGBA_ASTC_4x4x3_OES_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_4x4x4_OES:
+            return COMPRESSED_RGBA_ASTC_4x4x4_OES_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_5x4_KHR:
+            return COMPRESSED_RGBA_ASTC_5x4_KHR_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_5x4x4_OES:
+            return COMPRESSED_RGBA_ASTC_5x4x4_OES_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_5x5_KHR:
+            return COMPRESSED_RGBA_ASTC_5x5_KHR_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_5x5x4_OES:
+            return COMPRESSED_RGBA_ASTC_5x5x4_OES_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_5x5x5_OES:
+            return COMPRESSED_RGBA_ASTC_5x5x5_OES_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_6x5_KHR:
+            return COMPRESSED_RGBA_ASTC_6x5_KHR_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_6x5x5_OES:
+            return COMPRESSED_RGBA_ASTC_6x5x5_OES_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_6x6_KHR:
+            return COMPRESSED_RGBA_ASTC_6x6_KHR_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_6x6x5_OES:
+            return COMPRESSED_RGBA_ASTC_6x6x5_OES_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_6x6x6_OES:
+            return COMPRESSED_RGBA_ASTC_6x6x6_OES_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_8x5_KHR:
+            return COMPRESSED_RGBA_ASTC_8x5_KHR_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_8x6_KHR:
+            return COMPRESSED_RGBA_ASTC_8x6_KHR_to_default;
+        case GL_COMPRESSED_RGBA_ASTC_8x8_KHR:
+            return COMPRESSED_RGBA_ASTC_8x8_KHR_to_default;
         case GL_COMPRESSED_RGBA_BPTC_UNORM_EXT:
             return COMPRESSED_RGBA_BPTC_UNORM_EXT_to_default;
         case GL_COMPRESSED_RGBA_S3TC_DXT1_EXT:
@@ -1806,6 +3093,8 @@
         {
             switch (angleFormat)
             {
+                case FormatID::EAC_R11_SNORM_BLOCK:
+                    return COMPRESSED_SIGNED_R11_EAC_to_EAC_R11_SNORM_BLOCK;
                 case FormatID::R16_SNORM:
                     return COMPRESSED_SIGNED_R11_EAC_to_R16_SNORM;
                 default:
@@ -1817,6 +3106,8 @@
         {
             switch (angleFormat)
             {
+                case FormatID::EAC_R11G11_SNORM_BLOCK:
+                    return COMPRESSED_SIGNED_RG11_EAC_to_EAC_R11G11_SNORM_BLOCK;
                 case FormatID::R16G16_SNORM:
                     return COMPRESSED_SIGNED_RG11_EAC_to_R16G16_SNORM;
                 default:
@@ -1824,10 +3115,60 @@
             }
             break;
         }
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x10_KHR:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_10x10_KHR_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x5_KHR:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_10x5_KHR_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x6_KHR:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_10x6_KHR_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_10x8_KHR:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_10x8_KHR_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x10_KHR:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_12x10_KHR_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_12x12_KHR:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_12x12_KHR_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_3x3x3_OES:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_3x3x3_OES_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_4x3x3_OES:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_4x3x3_OES_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_4x4_KHR:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_4x4_KHR_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_4x4x3_OES:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_4x4x3_OES_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_4x4x4_OES:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_4x4x4_OES_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x4_KHR:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_5x4_KHR_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x4x4_OES:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_5x4x4_OES_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x5_KHR:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_5x5_KHR_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x5x4_OES:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_5x5x4_OES_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_5x5x5_OES:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_5x5x5_OES_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x5_KHR:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_6x5_KHR_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x5x5_OES:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_6x5x5_OES_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x6_KHR:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_6x6_KHR_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x6x5_OES:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_6x6x5_OES_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_6x6x6_OES:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_6x6x6_OES_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x5_KHR:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_8x5_KHR_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x6_KHR:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_8x6_KHR_to_default;
+        case GL_COMPRESSED_SRGB8_ALPHA8_ASTC_8x8_KHR:
+            return COMPRESSED_SRGB8_ALPHA8_ASTC_8x8_KHR_to_default;
         case GL_COMPRESSED_SRGB8_ALPHA8_ETC2_EAC:
         {
             switch (angleFormat)
             {
+                case FormatID::ETC2_R8G8B8A8_SRGB_BLOCK:
+                    return COMPRESSED_SRGB8_ALPHA8_ETC2_EAC_to_ETC2_R8G8B8A8_SRGB_BLOCK;
                 case FormatID::R8G8B8A8_UNORM_SRGB:
                     return COMPRESSED_SRGB8_ALPHA8_ETC2_EAC_to_R8G8B8A8_UNORM_SRGB;
                 default:
@@ -1839,6 +3180,8 @@
         {
             switch (angleFormat)
             {
+                case FormatID::ETC2_R8G8B8_SRGB_BLOCK:
+                    return COMPRESSED_SRGB8_ETC2_to_ETC2_R8G8B8_SRGB_BLOCK;
                 case FormatID::R8G8B8A8_UNORM_SRGB:
                     return COMPRESSED_SRGB8_ETC2_to_R8G8B8A8_UNORM_SRGB;
                 default:
@@ -1861,6 +3204,8 @@
         {
             switch (angleFormat)
             {
+                case FormatID::ETC2_R8G8B8A1_SRGB_BLOCK:
+                    return COMPRESSED_SRGB8_PUNCHTHROUGH_ALPHA1_ETC2_to_ETC2_R8G8B8A1_SRGB_BLOCK;
                 case FormatID::R8G8B8A8_UNORM_SRGB:
                     return COMPRESSED_SRGB8_PUNCHTHROUGH_ALPHA1_ETC2_to_R8G8B8A8_UNORM_SRGB;
                 default:
@@ -1895,6 +3240,8 @@
             {
                 case FormatID::D24_UNORM_S8_UINT:
                     return DEPTH24_STENCIL8_to_D24_UNORM_S8_UINT;
+                case FormatID::D32_FLOAT_S8X24_UINT:
+                    return DEPTH24_STENCIL8_to_D32_FLOAT_S8X24_UINT;
                 default:
                     break;
             }
@@ -1904,6 +3251,8 @@
         {
             switch (angleFormat)
             {
+                case FormatID::D24_UNORM_S8_UINT:
+                    return DEPTH32F_STENCIL8_to_D24_UNORM_S8_UINT;
                 case FormatID::D32_FLOAT_S8X24_UINT:
                     return DEPTH32F_STENCIL8_to_D32_FLOAT_S8X24_UINT;
                 default:
@@ -1928,6 +3277,8 @@
             {
                 case FormatID::D24_UNORM_S8_UINT:
                     return DEPTH_COMPONENT24_to_D24_UNORM_S8_UINT;
+                case FormatID::D32_FLOAT_S8X24_UINT:
+                    return DEPTH_COMPONENT24_to_D32_FLOAT_S8X24_UINT;
                 default:
                     break;
             }
@@ -1961,6 +3312,10 @@
         {
             switch (angleFormat)
             {
+                case FormatID::ETC1_R8G8B8_UNORM_BLOCK:
+                    return ETC1_RGB8_OES_to_ETC1_R8G8B8_UNORM_BLOCK;
+                case FormatID::ETC2_R8G8B8_UNORM_BLOCK:
+                    return ETC1_RGB8_OES_to_ETC2_R8G8B8_UNORM_BLOCK;
                 case FormatID::R8G8B8A8_UNORM:
                     return ETC1_RGB8_OES_to_R8G8B8A8_UNORM;
                 default:
@@ -2092,6 +3447,10 @@
             }
             break;
         }
+        case GL_R16_SSCALED_ANGLEX:
+            return R16_SSCALED_ANGLEX_to_default;
+        case GL_R16_USCALED_ANGLEX:
+            return R16_USCALED_ANGLEX_to_default;
         case GL_R32F:
         {
             switch (angleFormat)
@@ -2169,6 +3528,10 @@
             }
             break;
         }
+        case GL_R8_SSCALED_ANGLEX:
+            return R8_SSCALED_ANGLEX_to_default;
+        case GL_R8_USCALED_ANGLEX:
+            return R8_USCALED_ANGLEX_to_default;
         case GL_RG16F:
         {
             switch (angleFormat)
@@ -2224,6 +3587,10 @@
             }
             break;
         }
+        case GL_RG16_SSCALED_ANGLEX:
+            return RG16_SSCALED_ANGLEX_to_default;
+        case GL_RG16_USCALED_ANGLEX:
+            return RG16_USCALED_ANGLEX_to_default;
         case GL_RG32F:
         {
             switch (angleFormat)
@@ -2301,6 +3668,10 @@
             }
             break;
         }
+        case GL_RG8_SSCALED_ANGLEX:
+            return RG8_SSCALED_ANGLEX_to_default;
+        case GL_RG8_USCALED_ANGLEX:
+            return RG8_USCALED_ANGLEX_to_default;
         case GL_RGB:
             return RGB_to_default;
         case GL_RGB10_A2:
@@ -2325,12 +3696,40 @@
             }
             break;
         }
+        case GL_RGB10_A2_SINT_ANGLEX:
+        {
+            switch (angleFormat)
+            {
+                case FormatID::R10G10B10A2_SINT:
+                    return RGB10_A2_SINT_ANGLEX_to_R10G10B10A2_SINT;
+                default:
+                    break;
+            }
+            break;
+        }
+        case GL_RGB10_A2_SNORM_ANGLEX:
+        {
+            switch (angleFormat)
+            {
+                case FormatID::R10G10B10A2_SNORM:
+                    return RGB10_A2_SNORM_ANGLEX_to_R10G10B10A2_SNORM;
+                default:
+                    break;
+            }
+            break;
+        }
+        case GL_RGB10_A2_SSCALED_ANGLEX:
+            return RGB10_A2_SSCALED_ANGLEX_to_default;
+        case GL_RGB10_A2_USCALED_ANGLEX:
+            return RGB10_A2_USCALED_ANGLEX_to_default;
         case GL_RGB16F:
         {
             switch (angleFormat)
             {
                 case FormatID::R16G16B16A16_FLOAT:
                     return RGB16F_to_R16G16B16A16_FLOAT;
+                case FormatID::R16G16B16_FLOAT:
+                    return RGB16F_to_R16G16B16_FLOAT;
                 default:
                     break;
             }
@@ -2342,6 +3741,8 @@
             {
                 case FormatID::R16G16B16A16_SINT:
                     return RGB16I_to_R16G16B16A16_SINT;
+                case FormatID::R16G16B16_SINT:
+                    return RGB16I_to_R16G16B16_SINT;
                 default:
                     break;
             }
@@ -2353,6 +3754,8 @@
             {
                 case FormatID::R16G16B16A16_UINT:
                     return RGB16UI_to_R16G16B16A16_UINT;
+                case FormatID::R16G16B16_UINT:
+                    return RGB16UI_to_R16G16B16_UINT;
                 default:
                     break;
             }
@@ -2364,6 +3767,8 @@
             {
                 case FormatID::R16G16B16A16_UNORM:
                     return RGB16_EXT_to_R16G16B16A16_UNORM;
+                case FormatID::R16G16B16_UNORM:
+                    return RGB16_EXT_to_R16G16B16_UNORM;
                 default:
                     break;
             }
@@ -2375,17 +3780,25 @@
             {
                 case FormatID::R16G16B16A16_SNORM:
                     return RGB16_SNORM_EXT_to_R16G16B16A16_SNORM;
+                case FormatID::R16G16B16_SNORM:
+                    return RGB16_SNORM_EXT_to_R16G16B16_SNORM;
                 default:
                     break;
             }
             break;
         }
+        case GL_RGB16_SSCALED_ANGLEX:
+            return RGB16_SSCALED_ANGLEX_to_default;
+        case GL_RGB16_USCALED_ANGLEX:
+            return RGB16_USCALED_ANGLEX_to_default;
         case GL_RGB32F:
         {
             switch (angleFormat)
             {
                 case FormatID::R32G32B32A32_FLOAT:
                     return RGB32F_to_R32G32B32A32_FLOAT;
+                case FormatID::R32G32B32_FLOAT:
+                    return RGB32F_to_R32G32B32_FLOAT;
                 default:
                     break;
             }
@@ -2397,6 +3810,8 @@
             {
                 case FormatID::R32G32B32A32_SINT:
                     return RGB32I_to_R32G32B32A32_SINT;
+                case FormatID::R32G32B32_SINT:
+                    return RGB32I_to_R32G32B32_SINT;
                 default:
                     break;
             }
@@ -2408,6 +3823,8 @@
             {
                 case FormatID::R32G32B32A32_UINT:
                     return RGB32UI_to_R32G32B32A32_UINT;
+                case FormatID::R32G32B32_UINT:
+                    return RGB32UI_to_R32G32B32_UINT;
                 default:
                     break;
             }
@@ -2460,6 +3877,8 @@
             {
                 case FormatID::R8G8B8A8_SINT:
                     return RGB8I_to_R8G8B8A8_SINT;
+                case FormatID::R8G8B8_SINT:
+                    return RGB8I_to_R8G8B8_SINT;
                 default:
                     break;
             }
@@ -2471,6 +3890,8 @@
             {
                 case FormatID::R8G8B8A8_UINT:
                     return RGB8UI_to_R8G8B8A8_UINT;
+                case FormatID::R8G8B8_UINT:
+                    return RGB8UI_to_R8G8B8_UINT;
                 default:
                     break;
             }
@@ -2482,11 +3903,17 @@
             {
                 case FormatID::R8G8B8A8_SNORM:
                     return RGB8_SNORM_to_R8G8B8A8_SNORM;
+                case FormatID::R8G8B8_SNORM:
+                    return RGB8_SNORM_to_R8G8B8_SNORM;
                 default:
                     break;
             }
             break;
         }
+        case GL_RGB8_SSCALED_ANGLEX:
+            return RGB8_SSCALED_ANGLEX_to_default;
+        case GL_RGB8_USCALED_ANGLEX:
+            return RGB8_USCALED_ANGLEX_to_default;
         case GL_RGB9_E5:
         {
             switch (angleFormat)
@@ -2555,6 +3982,10 @@
             }
             break;
         }
+        case GL_RGBA16_SSCALED_ANGLEX:
+            return RGBA16_SSCALED_ANGLEX_to_default;
+        case GL_RGBA16_USCALED_ANGLEX:
+            return RGBA16_USCALED_ANGLEX_to_default;
         case GL_RGBA32F:
         {
             switch (angleFormat)
@@ -2645,6 +4076,10 @@
             }
             break;
         }
+        case GL_RGBA8_SSCALED_ANGLEX:
+            return RGBA8_SSCALED_ANGLEX_to_default;
+        case GL_RGBA8_USCALED_ANGLEX:
+            return RGBA8_USCALED_ANGLEX_to_default;
         case GL_SRGB8:
         {
             switch (angleFormat)
@@ -2674,6 +4109,7 @@
             break;
     }
     // clang-format on
+    ASSERT(internalFormat == GL_NONE || angleFormat == angle::FormatID::NONE);
     static LoadFunctionMap emptyLoadFunctionsMap;
     return emptyLoadFunctionsMap;
 